### ISSUE TITLE
Isolated declarations inconsistent about banning default assignment patterns for function parameters  

### ISSUE THREAD

**Issue by lucacasonato** [microsoft/TypeScript#58549]
### 🔎 Search Terms

- isolated declarations
- function parameter
- default value

### 🕗 Version & Regression Information

5.5.0-dev.20240514

### ⏯ Playground Link

https://www.typescriptlang.org/play/?target=99&isolatedDeclarations=true&ts=5.5.0-dev.20240514#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwJ4AUAhlFEXgFxwpIC2ARsFANoC6cAvHGwDRxoRaYEkRjQqxKjQZM+9KXUZQAlJwB81RU2VUAbhEwATOAG8AvgG4AUFdCRY-HAGd4ITqatwv+YqXIKZFnYuXkchETEoCSIApTlY7XVNQJ04fSNTMyszIA

### 💻 Code

```ts
export function y(array: number[] = [], comparator: (a: number, b: number) => number): void {};

export const x = {
    y(array: number[] = [], comparator: (a: number, b: number) => number): void {}
}
```

### 🙁 Actual behavior

For function `x` and `y` no diagnostic is raised on the property `array` at all.

For method `z.x`, a diagnostic is raised on `array` regarding the type annotation of `array` not including `undefined`. For method `z.y`, no diagnostic is raised.

### 🙂 Expected behavior

I would expect that in all four cases to raise an ID diagnostic on `array`, with the only possible fix being marking `comparator` as optional, or adding a default value assignment.

This is for multiple reasons:

1. The diagnostic raised for `z.x` suggests changing the type annotation to `T | undefined`. This is because without ID, TypeScript would automatically add `| undefined` to the emitted declarations. However, this is a bad suggestion, as doing so does not just change the external signature of the function, but also changes the type for the parameter of the function, in the function body, to a new type that does not match with the actual type of the parameter in the function (it can never be `undefined` in the body itself, because any `undefined` value would be replaced with the default argument).
2. An isolated declarations emitter can not determine whether it is safe to emit this or not, because knowing that requires knowing whether the type of `array` contains `undefined`. See

**Comment by dragomirtitian** [comment]
I opened #58637 This PR ensures we give errors for function declaration/class methods as well. It also only errors if the type doesn't syntactically contain undefined, and if the type can't be proven to not contain undefined. This means that `number[]` or `T | undefined` will not error, since for the first we know it does not contain `undefined` and for the second we know it does. The only case that will error is where the type of the parameter is a type reference such as `T` for which we would need to lookup type information to determine if it contains `undefined` or not.
